api/server.go

```diff
@@ -0,0 +1,75 @@
+package api
+
+import (
+	"context"
+	"net/http"
+	"time"
+
+	"github.com/julienschmidt/httprouter"
+	"github.com/segmentio/ksuid"
+	"go.uber.org/zap"
+)
+
+type server struct {
+	router *httprouter.Router
+	logger *zap.SugaredLogger
+	mw     []Middleware
+}
+
+// NewServer returns a HTTP server for accessing the the given API.
+func NewServer(addr string, logger *zap.SugaredLogger, a API) http.Server {
+	// Create our server, with default middlewares
+	s := server{
+		router: httprouter.New(),
+		logger: logger,
+		mw:     []Middleware{LogMW(logger), MetricsMW()},
```

Comment:
Should `MetricsMW` take `a.Endpoints()` as an arg so it can:
1) initialise the paths
2) not instrument paths that we don't support

With all the scrapers in the wild, `path` becomes kind of an unbounded set.

Probably also instantiate all potential `method` and `status`

Reply:
Yes, this is great. I will certainly do that. 

I was thinking about allowing `Options` to be passed to the server which will allow the user to configure the server in more ways, one of which I was thinking would be to choose paths to not be instrumented. 

I will check whether path could become unbounded, I am unsure whether this would happen or not.

Reply:
I have added this now @BdeUtra, the metrics are predefined, and look like this:

```
api_http_latency_seconds_bucket{method="GET",path="/accounts/:id",status="2XX",le="0.025"} 1
```

I have also tested a missing path; this does not currently get logged or added to any Prometheus metric. This is due to the default [NotFound Handler](https://godoc.org/github.com/julienschmidt/httprouter#Router.NotFound) in the httprouter library.

I should probably wrap that in a Prometheus metric (as well as the 'MethodNotAllowed' handler), to count how many 'bad' requests we are receiving, otherwise we will be blind to the true amount of traffic an API is receiving.

Reply:
Thanks dude!

Yeah that sounds good!

The canonical metric name would be `http_request_duration_seconds` if you want to follow that.